ultrasonic/src/main/kotlin/org/moire/ultrasonic/service/Downloader.kt

```diff
@@ -1,89 +1,107 @@
+ * TODO: Move entirely to subclass the Media3.DownloadService
  */
 class Downloader(
-    private val shufflePlayBuffer: ShufflePlayBuffer,
-    private val externalStorageMonitor: ExternalStorageMonitor,
-    private val localMediaPlayer: LocalMediaPlayer
+    private val storageMonitor: ExternalStorageMonitor,
+    private val legacyPlaylistManager: LegacyPlaylistManager,
 ) : KoinComponent {
 
-    private val playlist = mutableListOf<DownloadFile>()
+    // Dependencies
+    private val imageLoaderProvider: ImageLoaderProvider by inject()
+    private val activeServerProvider: ActiveServerProvider by inject()
+    private val mediaController: MediaPlayerController by inject()
 
     var started: Boolean = false
+    var shouldStop: Boolean = false
 
     private val downloadQueue = PriorityQueue<DownloadFile>()
     private val activelyDownloading = mutableListOf<DownloadFile>()
 
-    // TODO: The playlist is now published with RX, while the observableDownloads is using LiveData.
-    // Use the same for both
+    // The generic list models expect a LiveData, so even though we are using Rx for many events
+    // surrounding playback the list of Downloads is published as LiveData.
     val observableDownloads = MutableLiveData<List<DownloadFile>>()
 
-    private val jukeboxMediaPlayer: JukeboxMediaPlayer by inject()
-
     // This cache helps us to avoid creating duplicate DownloadFile instances when showing Entries
-    private val downloadFileCache = LRUCache<Track, DownloadFile>(100)
+    private val downloadFileCache = LRUCache<Track, DownloadFile>(500)
 
-    private var executorService: ScheduledExecutorService? = null
+    private var handler: Handler = Handler(Looper.getMainLooper())
     private var wifiLock: WifiManager.WifiLock? = null
 
-    private var playlistUpdateRevision: Long = 0
-        private set(value) {
-            field = value
-            RxBus.playlistPublisher.onNext(playlist)
-        }
+    private var backgroundPriorityCounter = 100
 
-    var backgroundPriorityCounter = 100
+    private val rxBusSubscription: CompositeDisposable = CompositeDisposable()
 
-    val downloadChecker = Runnable {
-        try {
-            Timber.w("Checking Downloads")
-            checkDownloadsInternal()
-        } catch (all: Exception) {
-            Timber.e(all, "checkDownloads() failed.")
+    init {
+        // Check downloads if the playlist changed
+        rxBusSubscription += RxBus.playlistObservable.subscribe {
+            checkDownloads()
+        }
+    }
+
+    private var downloadChecker = object : Runnable {
+        override fun run() {
+            try {
+                Timber.w("Checking Downloads")
+                checkDownloadsInternal()
+            } catch (all: Exception) {
+                Timber.e(all, "checkDownloads() failed.")
+            } finally {
+                if (!shouldStop) {
+                    Handler(Looper.getMainLooper()).postDelayed(this, CHECK_INTERVAL)
```

Comment:
For me it seems every time the downloadChecker is started, it will create a loop of itself which checks every five seconds. In a short time there will be tens of downloadChecker instances which will all check in every five seconds, flooding everything.
I think if checking in every five seconds is necessary, there should be two checkers: one that is run every five seconds, and one that can be started ad-hoc, and will run only once.

Reply:
This should not happen!
We should keep a reference to the runnable and if it's active don't create a new one....